x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/sender/RequestExecutorService.java

```diff
@@ -221,7 +259,83 @@ private void scheduleNextHandleTasks(TimeValue timeToWait) {
+    }
+
+    private void processRequestQueue() {
+        try {
+            while (isShutdown() == false) {
+                // Blocks the request queue thread until a new request comes in
+                var task = (RequestTask) requestQueue.take();
+
+                if (isShutdown()) {
+                    logger.debug("Shutdown requested while handling request tasks, cleaning up");
+                    cleanup();
+                    return;
+                } else {
+                    var requestManager = task.getRequestManager();
+
+                    if (rateLimitingEnabled(requestManager)) {
+                        submitTaskToRateLimitedExecutionPath(task);
+                    } else {
+                        executeTaskImmediately(task);
+                    }
+                }
+            }
+        } catch (InterruptedException e) {
+            // Restore interrupt to propagate to the calling thread
+            Thread.currentThread().interrupt();
+            logger.debug("Inference request queue interrupted, exiting");
+        } catch (Exception e) {
+            logger.warn("Error processing task in inference request queue", e);
+            cleanup();
+        }
+    }
+
+    private void executeTaskImmediately(RequestTask task) {
+        try {
+            task.getRequestManager()
+                .execute(task.getInferenceInputs(), requestSender, task.getRequestCompletedFunction(), task.getListener());
+        } catch (Exception e) {
+            logger.warn(
+                format("Failed to execute fast-path request for inference id [%s]", task.getRequestManager().inferenceEntityId()),
+                e
+            );
+
+            task.onRejection(
+                new EsRejectedExecutionException(
+                    format("Failed to execute request for inference id [%s]", task.getRequestManager().inferenceEntityId()),
+                    false
+                )
+            );
+        }
+    }
+
+    // visible for testing
+    void submitTaskToRateLimitedExecutionPath(RequestTask task) {
+        var requestManager = task.getRequestManager();
+        var endpoint = rateLimitGroupings.computeIfAbsent(requestManager.rateLimitGrouping(), key -> {
+            var endpointHandler = new RateLimitingEndpointHandler(
+                Integer.toString(requestManager.rateLimitGrouping().hashCode()),
+                queueCreator,
+                settings,
+                requestSender,
+                clock,
+                requestManager.rateLimitSettings(),
+                this::isShutdown,
+                rateLimiterCreator,
+                rateLimitDivisor.get()
+            );
+
+            endpointHandler.init();
+            return endpointHandler;
+        });
+
+        endpoint.enqueue(task);
+    }
+
+    private boolean rateLimitingEnabled(RequestManager requestManager) {
```

Comment:
For consistency how about we make this static and accept a `RateLimitSettings` object. Then we can use it in `executeEnqueuedTaskInternal` which has a similar check. Technically `executeEnqueuedTaskInternal` should never receive a non-rate limited task, but it'd probably be good to check just in case.

Reply:
Good point, adjusted with [Reuse rateLimitSettingsEnabled check](https://github.com/elastic/elasticsearch/pull/136167/commits/174526cb314142d1158aa9d4a0c8c1f2f980ed4c)